data_structures/disjoint_set/disjoint_set.ts

```diff
@@ -0,0 +1,71 @@
+/**
+ * A Disjoint Set is a data structure that keeps track of a set of elements
+ * partitioned into a number of disjoint (non-overlapping) subsets.
+ * Elements are uniquely represented by an index (0-based).
+ * 
+ * It uses union by size: The smaller set is joined to the bigger one.
+ * This allows to keep the `find` complexity to O(logn).
+ * 
+ * You can perform the following operations on the disjoint set:
+ * - find: Determine which subset a particular element is in - O(logn) ~ O(1) with path compression.
+ * - join: Join two subsets into a single subset - O(1).
+ * - isSame: Check if two elements are in the same subset - O(1).
+ */
+export class DisjointSet {
+  /** Direct parent for an element */
+  private head: number[];
+
+  /** Size of the subtree above an element */
+  private size: number[];
+
+  constructor(n: number) {
+    // Initially each set has it's own id element
+    this.head = [...new Array(n)].map((_, index) => index);
+    this.size = new Array(n).fill(1);
+  }
+
+  /**
+   * Find the representative index for an element
+   */
+  find(index: number): number {
+    if (this.head[index] != index) {
+      // Use path compression (Set an edge between the element and it's head)
+      this.head[index] = this.find(this.head[index]);
+    }
+    return this.head[index];
+  }
+
+
+  /**
+   * Join two sets
+   * It uses union by size to leverage `find` complexity to O(logn)
```

Comment:
This comment makes no sense. `find` complexity is supposedly O(1) due to path compression, as should join complexity be according to your comment at the top.